src/main/kotlin/com/fasterxml/jackson/module/kotlin/ReflectionCache.kt

```diff
@@ -49,6 +52,37 @@ internal class ReflectionCache(reflectionCacheSize: Int) {
     fun kotlinFromJava(key: Method): KFunction<*>? = javaMethodToKotlin.get(key)
             ?: key.kotlinFunction?.let { javaMethodToKotlin.putIfAbsent(key, it) ?: it }
 
+    /**
+     * return null if...
+     * - can't get kotlinFunction
+     * - contains extensionReceiverParameter
+     * - instance parameter is not companion object or can't get
+     */
+    @Suppress("UNCHECKED_CAST")
+    fun valueCreatorFromJava(_withArgsCreator: AnnotatedWithParams): ValueCreator<*>? = when (_withArgsCreator) {
+        is AnnotatedConstructor -> {
+            val constructor = _withArgsCreator.annotated as Constructor<Any>
+
+            javaConstructorToValueCreator.get(constructor)
+                ?: kotlinFromJava(constructor)?.let {
+                    val value = ConstructorValueCreator(it)
+
+                    javaConstructorToValueCreator.putIfAbsent(constructor, value) ?: value
+                }
+        }
+        is AnnotatedMethod -> {
+            val method = _withArgsCreator.annotated as Method
+
+            javaMethodToValueCreator.get(method)
+                ?: kotlinFromJava(method)?.let {
+                    val value = MethodValueCreator.of(it)
+
+                    javaMethodToValueCreator.putIfAbsent(method, value) ?: value
+                }
+        }
+        else -> throw IllegalStateException("Expected a constructor or method to create a Kotlin object, instead found ${_withArgsCreator.annotated.javaClass.name}")
+    } // we cannot reflect this method so do the default Java-ish behavior
```

Comment:
Won't the `else -> throw` prevent any further processing? Perhaps it should be `else -> null`?

Reply:
The reason is that the following process has been ported as is.
https://github.com/FasterXML/jackson-module-kotlin/blob/6023113a10f07ebf087eb767c162ae904b5f7149/src/main/kotlin/com/fasterxml/jackson/module/kotlin/KotlinValueInstantiator.kt#L40-L48

Reply:
Got it. For posterity the null half of that `?:` was moved to [`KotlinValueInstantiator:31`](https://github.com/FasterXML/jackson-module-kotlin/pull/512/files/3ffe0a0f1de0284c3fb7ff8138f6fe711cde8853#diff-06113cbaefa75226635da3db472eb87b911d3584948a17ab3b4a70daaf13d311R31)